### ISSUE TITLE
@ConfigurationProperties are not available in conditions

### ISSUE THREAD

**Issue by hekonsek** [spring-projects/spring-boot#2282]
Hi,

There is a general issue with `@ConfigurationProperties` and beans/classes conditions not working well together. Basically `@ConfigurationProperties` properties are instantiated *after* conditions are evaluated. As a result, conditions can't see those properties.

For example creating `@ConfigurationProperties` with some default true value...

    @ConfigurationProperties(prefix = "foo")
    public class MyConfigurationProperties {
    
      boolean bar = true;
    
      // getters & setters
    
    ...
    
    }

...and using it in `@ConditionalOnProperty("foo.bar")` doesn't evaluate to `true` (because `OnPropertyCondition#getMatchOutcome` is executed before `MyConfigurationProperties` is processed by the `ConfigurationClassPostProcessor`, so `PropertyResolver` can't see `foo.bar` property yet).

The problem is not limited to `@ConditionalOnProperty`. It affects all conditions that relies on properties *or* on the other beans referring `@ConfigurationProperties`.

Fix to that will be highly appreciated :) .

Cheers.

**Comment by snicoll** [comment]
Well, the link is reverse. The `Environment` injects value into `@ConfigurationProperties` annotated bean, not the other way around. So there is no way (currently) that the default value of your `bar` property would **update** the `Environment`.

**Comment by hekonsek** [comment]
Is there a chance to respect `@ConfigurationProperties` default values? Default values are pretty common in `@ConfigurationProperties` instances, so this is quite surprising for end-users that `@ConditionalOnProperty("foo.bar")` (and `PropertyResolver` in general) can't see them.

BTW The another painful scenario is when we define bean referring `@ConfigurationProperties` instance...

    class MyBean {
      @Autowired
      MyProperties properties;
    
      void readFoo(){
        assertEquals("fooValue", properties.getFoo());
      }
    }

If I access `MyBean#readFoo` using `ConditionContext`'s bean factory, the value of `foo` property will not be properly evaluated (because environment has not been bound to the `MyProperties` at the moment of the condition evaluation).

In general with these issues in place, using `@Value("${foo:default}")` instead of `@ConfigurationProperties` sounds more predictable and intuitive.

Do you think that this kind of binding corner cases could be addressed? I (and not only me I guess) will highly appreciate it.

**Comment by philwebb** [comment]
It's probably going to be pretty hard to make `@ConditionalOnProperty` look at the default values of the `@ConfigurationProperties`. As you said, currently the bind occurs after the bean has been created.

Have you seen the `matchIfMissing` and `havingValue` attributes of `@ConditionalOnProperty`. We use those internally quite a bit to do something similar to your first example:

```java
@ConfigurationProperties(prefix = "foo")
@ConditionalOnProperty(prefix="foo", name="bar", matchIfMissing=true)
public class MyConfigurationProperties {

    private boolean bar = true;
    
    // getters & setters

}
```

**Comment by hekonsek** [comment]
Hi Phil,

Yes, actually I use `matchIfMissing` as the workaround for cases like this. :) However this is a bit ugly, as I really need to duplicate default value (keep in mind that `matchIfMissing=true` is really the duplication of the default `@ConfigurationProperty`  like `bar = true`). And If I change default value in `@ConfigurationPropeties` I need to remember to change `matchIfMissing` in the appropriate bean/config. This workaround also works only for booleans - default String value won't work:

      @ConditionalOnProperty(prefix="foo", name="bar", havingValue="defaultStringValue")

My main concern here is that currently it is counter-intuitive that `@ConditionalOnProperty` or bean accessed via `ConditionContext` don't respect `@ConfigurationProperties` contract. I personally can live with that, but this is issue that will confuse some users or even sneak some hard-to-spot bugs into their production environments. I love Spring Boot and I recommend it to all the users/customers I work with, so I just try to predict some issues that may occur :) .

Can you consider keeping this issue open? I understand that this can be a tricky beast but I believe that it is worth some investigation in the future.

Laters!

**Comment by philwebb** [comment]
Yeah, lets keep this open and take a look when the 1.2 bugs die down. It might be possible to update `@ConditionalOnProperty` to direct inspect the class that it's on to see if it is a `@ConfigurationProperties`. I'm not very keen to couple those classes though.

**Comment by snicoll** [comment]
you could also read the meta-data at runtime. We're already doing that for the `configprops` endpoint.

**Comment by hekonsek** [comment]
Thanks guys.

**Comment by longfeizheng** [comment]
`
@configuration
@enableconfigurationproperties(ServiceProps .class)
public class CoreConfig {

}
`

**Comment by snicoll** [comment]
@longfeizheng could you please pasting the same comment in  several issues? If you have a question, please ask on StackOverflow or join us on Gitter.

**Comment by IdanE** [comment]
If anyone is facing this issue, I've written a [post processor](https://gist.github.com/IdanE/84e7aed06a29918100f30400270b65d4)  to inject defaults into the environment

**Comment by jamesmehorter** [comment]
👋 Hi all, Seems I may have just stumbled into this issue as well? Please correct me if I'm mistaken

```
// application.properties
foo.bar=raboof

// Config.kt
@ConfigurationProperties(prefix = "foo")
data class Config {
    var bar: String? = null
}

// Bean.kt
@Bean
@Conditional(hasBar::class)
fun something( ... ) {
    ...
}

// Condition.kt
@EnableConfigurationProperties(Config::class)
class hasBar(
    //val config: Config // "No argument provided for a required parameter"
) : Condition {
    override fun matches(context: ConditionContext, ...) {
        val config = context.beanFactory?.getBean('Config') as Config
        println(config) // outputs Config[bar=null]

        println(context.environment.getProperty("foo.bar")) // outputs raboof
        ...
    }
}
```

**Comment by wilkinsona** [comment]
@jamesmehorter You can't safely retrieve beans from the context during condition evaluation as they are called too early in the application context's lifecycle.

**Comment by Edhendil** [comment]
While `@ConfigurationProperties` are not available during `@Conditional` annotations execution you can still convert properties into POJOs with the use of `Binder` class. This means that properties will be bound to configuration classes mutliple times during application startup but it's a relatively small price to pay for being able to execute any logic on mapped properties.

Just create a custom conditional implementation and use this line.

```
Object bean = Binder.get(context.getEnvironment())
    .bind("your.properties.prefix", YourConfigurationPropertiesClass.class).orElse(null);
```

Here's a [gist with full solution](https://gist.github.com/Edhendil/9b52bc936ae459c31619b6c89ac5acac).

**Comment by michaelzangl** [comment]
@tokrug The gist does not work any more. Do you still have it somewhere?

**Comment by tokrug** [comment]
@michaelzangl I have updated the link in the original comment.

**Comment by albertocavalcante** [comment]
Is this use case being considered to be first-class supported in Spring Boot? Thanks

**Comment by philwebb** [comment]
@albertocavalcante The issue is open an in the "General Backlog" milestone which means we're still considering it. Having said that, this we have a number of other tasks that are consuming a lot of our bandwidth so we're very unlikely to get to this one anytime soon.

**Comment by rajjaiswalsaumya** [comment]
@philwebb Can you please revisit this and consider this one?

**Comment by philwebb** [comment]
@rajjaiswalsaumya I'm afraid the comment just above yours still stands. This isn't a particularly high priority compared to the other work we have on.

**Comment by hantsy** [comment]
In my case I used a `Map<String, Any>` as the properties to make the key accept dynamic values. 

```kotlin
@ConfigurationProperties(perfix="test")
data class {
  props:Map<String, Any>
}

``

So the `@ConditionalOnProperty("test.props")` or `@ConditionalOnProperty(perfix= "test" name=["props"])` does not work in my case.

**Comment by mauromol** [comment]
Just my additional 2 cents on this. Apart from `@ConditionalOnProperty`, one more general problem I discovered today (that is also indirectly mentioned here) is that `@Value` or property placeholders in properties are not resolved if you want to reference another property that is not _explicitly_ defined in a properties file. That was a huge "WTF" for me today, even if I have been using Spring Boot since many years now :-D

Concrete example: as documented in the official documentation, [server.tomcat.max-connections](https://docs.spring.io/spring-boot/appendix/application-properties/index.html#application-properties.server.server.tomcat.max-connections) has a default value of 8192. So I feel I am allowed to reference `server.tomcat.max-connections` like this in a Java class:

```java
  @Value("${server.tomcat.max-connections}") int myValue;
```

or like this in `application.properties`:

```properties
my-other-prop=${server.tomcat.max-connections}
```

However, since this default value of 8192 is actually defined just in the `ServerProperties.Tomcat` Java source, this does not work at all _unless_ I set an explicit value for it anywhere in my properties file(s). (BTW: yes, of course, this was observed in an application using the web starter to run an embedded Tomcat, so Tomcat properties are registered and enabled in the context)

I understand this is due to what @snicoll said:
> The Environment injects values into @ConfigurationProperties annotated bean, not the other way around.

However, from a practical point of view, this is quite unexpected and unnatural IMHO.
At the very least, I would highlight this when talking about the differences between `@Value` and `ConfigurationProperties`:
https://docs.spring.io/spring-boot/reference/features/external-config.html#features.external-config.typesafe-configuration-properties.vs-value-annotation
or when talking about property placeholders at https://docs.spring.io/spring-boot/reference/features/external-config.html#features.external-config.files.property-placeholders.

**Comment by snicoll** [comment]
>  this is quite unexpected and unnatural IMHO

What is "this"? `8192` is [set in code](https://github.com/spring-projects/spring-boot/blob/dc54595f7a99ed93ef72f7f50e73538a8c85c73d/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java#L457), why would you want a property that isn't defined anywhere to have that value? 

> differences between @Value and ConfigurationProperties

Your own comment shows that there isn't a difference between `@Value` and configuring the environment via application.properties, which is exactly what [that table describes](https://docs.spring.io/spring-boot/reference/features/external-config.html#features.external-config.typesafe-configuration-properties.vs-value-annotation).

**Comment by mauromol** [comment]
> > this is quite unexpected and unnatural IMHO
> 
> What is "this"? `8192` is [set in code](https://github.com/spring-projects/spring-boot/blob/dc54595f7a99ed93ef72f7f50e73538a8c85c73d/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ServerProperties.java#L457), why would you want a property that isn't defined anywhere to have that value?

If the appendix says that the default value for `server.tomcat.max-connections` is 8192, I expect that any reference to `server.tomcat.max-connections` property evaluates to 8192, if not otherwise overridden. Instead, this is not true, unless you get this value through an instance of `ServerProperties`. That's what feels unexpected to me. The fact that 8192 is specified in code and not in a property source should be an implementation detail IMHO. Unless this is stated clearly. If it is, please apologise, but I couldn't find it mentioned anywhere in documentation. I just read that ConfigurationProperties are a convenient way to handle externalised configuration in a more structured and type-safe way.

> Your own comment shows that there isn't a difference between `@Value` and configuring the environment via application.properties, which is exactly what [that table describes](https://docs.spring.io/spring-boot/reference/features/external-config.html#features.external-config.typesafe-configuration-properties.vs-value-annotation).

Hmmm... that table is in a section that is named "ConfigurationProperties vs `@Value`", and highlights differences in behaviour when you use one mechanism or the other. What I'm describing there falls exactly in this category. If you try to get the above property through a `ServerProperties` you get 8192, if you do the same with `@Value` you get an unresolved string instead. Unless you give an explicit value to that property, in which case then you start getting the same value instead. Same applies to placeholders used to define the value of other properties.

**Comment by wilkinsona** [comment]
We can clarify the docs here as it's clear from the recent comments that there's room for improvement.

The documentation is written from the perspective of someone configuring a Spring Boot application, not someone trying to use one of Spring Boot's properties for their own purposes. The latter is far less common than the former but I think that it's still worth mentioning in the docs.

There's no behavioral difference in terms of default values between your own code that uses `@Value` and your own code that uses `@ConfigurationProperties`. In either case, you won't get the default value that's defined in Spring Boot's own `@ConfigurationProperties` class. It's this that we need to clarify. I've opened https://github.com/spring-projects/spring-boot/issues/49654.

**Comment by quaff** [comment]
I'm wondering if we could append a synthesized `PropertySource` which including all default values, then we can safely use `@Value("${server.tomcat.max-connections}")`.

**Comment by wilkinsona** [comment]
I don't think we should do that as it would have a sizeable impact on startup time. We'd have to scan the entire classpath for configuration property metadata, parse that json and extract all of the properties and their default values and then populate the property source. Possibly more significantly, we'd then have far more properties to bind which will also slow things down. Right now, we can shortcut the binding process if there are no properties to be bound to a particular target (nested or top-level). This optimisation would be undone for any target with defaults.

**Comment by quaff** [comment]
> Possibly more significantly, we'd then have far more properties to bind which will also slow things down. Right now, we can shortcut the binding process if there are no properties to be bound to a particular target (nested or top-level). This optimisation would be undone for any target with defaults.

We could exclude this special `PropertySource` for binding process, it's used for `env.getProperty()`, then we could use default properties for `@ConfigurationProperties` and `@Value`.

**Comment by wilkinsona** [comment]
We discussed this today and wondered about having something like this:

```java
@ConditionalOnBoundProperty(class = ExampleProperties.class, property="flag")
```

It would bind `ExampleProperties` and then use reflection to call `getFlag()` (JavaBean) or `flag()` (record) to get the value. 

We're not yet convinced that it's a good idea.

The subtle difference between `@ConditionalOnProperty` and `@ConditionalOnBoundProperty` may be confusing. It also wouldn't if the actual `@ConfigurationProperties` bean that's created after condition evaluation has a custom default value:

```java
@Bean
ExampleProperties exampleProperties() {
    return new ExampleProperties(false);
}
```

`@Name` also complicates things further. Is the `property` attribute referencing the property or the method to call? If it's the former (which is perhaps the more intuitive), we'd have to do some introspection to find the right getter to call. There's code for this in the binder, but it's an implementation detail and not intended for re-use.

We also haven't needed it in Boot, despite using `@ConditionalOnProperty` fairly broadly. Generally speaking, we don't have both `@ConditionalOnProperty` and `@ConfigurationProperties` class dealing with the same property so there's nothing to keep in sync.

**Comment by philwebb** [comment]
We discussed this today and we've decided this isn't something we want to do. Thanks anyway for the suggestions and discussion.
